hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizerOnCluster.java

```diff
@@ -179,6 +182,71 @@ void testRegionNormalizationSplitOnCluster(boolean limitedByQuota) throws Except
     admin.deleteTable(TABLENAME);
   }
 
+  // This test is to make sure that normalizer is only going to merge adjacent regions.
+  @Test
+  @SuppressWarnings("deprecation")
+  public void testNormalizerCannotMergeNonAdjacentRegions() throws Exception {
+    final TableName tableName = TableName.valueOf(name.getMethodName());
+    MiniHBaseCluster cluster = TEST_UTIL.getHBaseCluster();
+    HMaster m = cluster.getMaster();
+
+    // create 5 regions with sizes to trigger merge of small regions
+    final byte[][] keys = {
+      Bytes.toBytes("aa"),
+      Bytes.toBytes("aa1"),
+      Bytes.toBytes("aa1!"),
+      Bytes.toBytes("aa2")
+    };
+
+    try (Table ht = TEST_UTIL.createMultiRegionTable(tableName, new byte[][]{FAMILYNAME}, keys)) {
+      // Need to get sorted list of regions here, the order is
+      // [, "aa"), ["aa", "aa1"), ["aa1", "aa1!"), ["aa1!", "aa2"), ["aa2", )
+      List<HRegion> generatedRegions = TEST_UTIL.getHBaseCluster().getRegions(tableName);
+      Collections.sort(generatedRegions, Comparator.comparing(HRegion::getRegionInfo, RegionInfo.COMPARATOR));
+
+      // Region ["aa", "aa1") and ["aa1!", "aa2") are not adjacent, they are not supposed to
+      // merged.
+      HRegion region = generatedRegions.get(0);
+      generateTestData(region, 3);
+      region.flush(true);
+
+      region = generatedRegions.get(1);
+      generateTestData(region, 1);
+      region.flush(true);
+
+      region = generatedRegions.get(2);
+      generateTestData(region, 3);
+      region.flush(true);
+
+      region = generatedRegions.get(3);
+      generateTestData(region, 1);
+      region.flush(true);
+
+      region = generatedRegions.get(4);
+      generateTestData(region, 5);
+      region.flush(true);
+
+      ModifyableTableDescriptor htd = new ModifyableTableDescriptor(tableName,
+        admin.getDescriptor(tableName));
+
+      htd.setNormalizationEnabled(true);
+      admin.modifyTable(htd);
+
+      admin.flush(tableName);
+
+      assertEquals(5, MetaTableAccessor.getRegionCount(TEST_UTIL.getConnection(), tableName));
+
+      Thread.sleep(5000); // to let region load to update
```

Comment:
Is there a condition you can `waitFor` instead of a fixed sleep? These fixed sleeps lead to slow and flakey tests :(

Reply:
Let me check one, it is just copy and paste here from another test.

Reply:
Understood.

Reply:
I went through the code. The logic is like this, every 3 seconds (default config), region servers send sever metrics to master and update the copy in master in-memory database, which is used by normalizer to calculate region sizes. We can (not) manually trigger a server metrics update from region server as the method is a protected one for testing in regionserver module. Even if it does a manual server metrics update from region server, it still needs to wait(sleep) until master finishes rpc processing. Will leave as it is unless there is something missing from my code reading, thanks.

Reply:
Maybe we can do manual server metrics update followed by sleep and then assert null plan?

Reply:
The manual server mtetrics update will need to expose the protected methods out of the module, and after that, there is still no way to guarantee the latest update is processed by master (sleep is still needed). I thought about it and gave up. It is mostly from testRegionNormalizationMergeOnCluster() and it never shows up in flaky test list, so it is ok here.